unlimited_channel.go

```diff
@@ -8,103 +8,112 @@ import (
 
 var ErrEventQueueClosed = errors.New("zk: event queue closed")
 
-type EventQueue interface {
-	// Next waits for a new Event to be received until the context expires or the queue is closed.
-	Next(ctx context.Context) (Event, error)
-	// Push adds the given event to the queue and notifies any in-flight calls to Next that a new event is available.
-	Push(e Event)
-	// Close functions like closing a channel. Subsequent calls to Next will drain whatever events remain in the buffer
-	// while subsequent calls to Push will panic. Once remaining events are drained, Next will return
+type Queue[T any] interface {
+	// Next waits for a new element to be received until the context expires or the queue is closed.
+	Next(ctx context.Context) (T, error)
+	// Push adds the given element to the queue and notifies any in-flight calls to Next that a new element is
+	// available.
+	Push(e T)
+	// Close functions like closing a channel. Subsequent calls to Next will drain whatever elements remain in the
+	// buffer while subsequent calls to Push will panic. Once remaining elements are drained, Next will return
 	// ErrEventQueueClosed.
 	Close()
 }
 
-type chanEventQueue chan Event
+// EventQueue is added to preserve the old EventQueue type which had an equivalent interface, for backwards
+// compatibility in method signatures.
+type EventQueue = Queue[Event]
 
-func (c chanEventQueue) Next(ctx context.Context) (Event, error) {
+type ChanQueue[T any] chan T
+
+func (c ChanQueue[T]) Next(ctx context.Context) (T, error) {
 	select {
 	case <-ctx.Done():
-		return Event{}, ctx.Err()
+		var t T
```

Comment:
Why do we need to declare `T` before returning? Can you not return `nil` in this case because `T` isn't nullable or something?

Reply:
Yeah it's the only way to get a 0-value of type `T`. Pretty frustrating